File: app/src/main/java/com/tichuguru/db/TichuDatabase.java
package com.tichuguru.db;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

@Database(entities = {PlayerEntity.class, GameEntity.class, HandEntity.class}, version = 1, exportSchema = false)
public abstract class TichuDatabase extends RoomDatabase {
    private static volatile TichuDatabase instance;

    @NonNull public abstract PlayerDao playerDao();
    @NonNull public abstract GameDao gameDao();
    @NonNull public abstract HandDao handDao();

    @NonNull
    public static TichuDatabase getInstance(@NonNull Context context) {
        if (instance == null) {
            synchronized (TichuDatabase.class) {
                if (instance == null) {
                    instance = Room.databaseBuilder(context.getApplicationContext(), TichuDatabase.class, "tichu.db")
                            .allowMainThreadQueries()
                            .build();
                }
            }
        }
        return instance;
    }
}

File: app/src/main/kotlin/com/tichuguru/db/TichuDatabase.kt
package com.tichuguru.db

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase

@Database(entities = [PlayerEntity::class, GameEntity::class, HandEntity::class], version = 1, exportSchema = false)
abstract class TichuDatabase : RoomDatabase() {

    abstract fun playerDao(): PlayerDao
    abstract fun gameDao(): GameDao
    abstract fun handDao(): HandDao

    companion object {
        @Volatile private var instance: TichuDatabase? = null

        @JvmStatic
        fun getInstance(context: Context): TichuDatabase =
            instance ?: synchronized(this) {
                instance ?: Room.databaseBuilder(context.applicationContext, TichuDatabase::class.java, "tichu.db")
                    .allowMainThreadQueries()
                    .build()
                    .also { instance = it }
            }
    }
}
